File: Python/src/util/benchmark/stopwatch.py
"""
A stopwatch for timing laps, ported from util/benchmark/Stopwatch.java.

The Java implements AutoCloseable, so it is used with try-with-resources; here
it is a context manager, used with ``with``.
"""

from __future__ import annotations

import time

MILLION = 1_000_000

_TIME_FACTORS = {
    "milliseconds": MILLION,
    "seconds": 1_000 * MILLION,
    "microseconds": 1_000,
    "nanoseconds": 1,
}


class Stopwatch:
    """
    Time successive laps, in units fixed when the Stopwatch is created.

    Each call to ``lap`` reports the time since the previous call (or since
    construction) and starts the next lap.
    """

    def __init__(self, units: str | None = "milliseconds") -> None:
        """
        :param units: one of "nanoseconds", "microseconds", "milliseconds" or
                      "seconds". None means milliseconds.
        :raises ValueError: if units is none of those.
        """
        self._time_factor = _calculate_time_factor(units)
        self._start: int | None = _clock()

    def lap(self) -> int:
        """
        End the current lap and begin the next.

        :return: the duration of the lap just ended, in this Stopwatch's units.
        """
        assert self._start is not None, "Stopwatch is closed"
        lap_start = self._start
        self._start = _clock()
        # NOTE each reading is divided before the subtraction, following the
        # Java. That is not the same as dividing the difference: it can be off by
        # one unit, because the two readings truncate independently.
        return self._start // self._time_factor - lap_start // self._time_factor

    def close(self) -> None:
        """
        Close this Stopwatch. Calling ``lap`` afterwards is an error.
        """
        self._start = None

    def __enter__(self) -> Stopwatch:
        return self

    def __exit__(self, exc_type: object, exc_value: object, traceback: object) -> None:
        self.close()


def _calculate_time_factor(units: str | None) -> int:
    """
    :param units: the name of the units.
    :return: the number of nanoseconds in one of those units.
    :raises ValueError: if the units are not recognized.
    """
    if units is None:
        return MILLION
    factor = _TIME_FACTORS.get(units)
    if factor is None:
        raise ValueError(f"Invalid time units: {units}")
    return factor


def _clock() -> int:
    """
    :return: the reading of a monotonic clock, in nanoseconds. Monotonic because
             the wall clock can move backwards, which would give a negative lap.
    """
    return time.monotonic_ns()

File: Java/src/main/java/com/phasmidsoftware/dsaipg/util/benchmark/Stopwatch.java
package com.phasmidsoftware.dsaipg.util.benchmark;

/**
 * Simple benchmarking tool: Stopwatch.
 * There is no warm-up here, no pause/resume and no repetition (for these use Timer).
 * It is simply a convenient way to time an execution with results in milliseconds (the default),
 * microseconds, nanoseconds, or seconds.
 * <p>
 * Once the Stopwatch is started (i.e. constructed), you "read" the stopwatch by calling lap(),
 * which returns the number of time units since the previous lap (or since the start).
 * <p>NOTE that Stopwatch is not intended for long laps because it lacks the required precision
 * unless you use seconds as the time unit.</p>
 */
public class Stopwatch implements AutoCloseable {

    /**
     * Constructs a Stopwatch instance with a specified time factor to determine the time unit
     * used for measuring elapsed time.
     *
     * @param timeFactor the factor that defines the granularity of time measurement;
     *                   corresponds to the number of nanoseconds in one time unit.
     */
    private Stopwatch(int timeFactor) {
        this.timeFactor = timeFactor;
        start = System.nanoTime();
    }

    /**
     * Constructs a Stopwatch instance initialized with the specified time units.
     * The time units determine the granularity for measuring elapsed time.
     *
     * @param units the name of the time unit to be used for measurement, e.g., "milliseconds",
     *              "seconds", or "microseconds". If an invalid or unsupported unit is supplied,
     *              an IllegalArgumentException is thrown.
     */
    public Stopwatch(String units) {
        this (calculateTimeFactor(units));
    }

    /**
     * Construct and start a Stopwatch which yields times in milliseconds.
     */
    public Stopwatch() {
        this("milliseconds");
    }

    /**
     * @return the number of time units elapsed since the last lap (or start).
     * @throws AssertionError if this Stopwatch has been closed already.
     */
    public long lap() {
        assert start != null : "Stopwatch is closed";
        final long lapStart = start;
        start = System.nanoTime();
        // NOTE divide the elapsed time, not each reading. Dividing the readings
        // separately truncates each one independently, so the result could be
        // out by one unit in either direction.
        return (start - lapStart) / timeFactor;
    }

    /**
     * Close this Stopwatch.
     * NOTE that you should run the Stopwatch within a try-with-resource statement. See unit tests.
     */
    public void close() {
        start = null;
    }

    static final int MILLION = 1_000_000;

    /**
     * Determines the time factor based on the provided time unit.
     * The time factor is used to adjust time measurements to the specified unit.
     *
     * @param units the name of the time unit, which can be "milliseconds", "seconds",
     *              "microseconds", or "nanoseconds". If null or "milliseconds" is provided,
     *              the factor corresponds to one million nanoseconds.
     * @return an integer value representing the time factor for the specified unit.
     * @throws IllegalArgumentException if the provided time unit is invalid or unsupported.
     */
    private static int calculateTimeFactor(String units) {
        if (units == null || units.equals("milliseconds"))
            return MILLION;
        else if (units.equals("seconds"))
            return 1_000 * MILLION;
        else if (units.equals("microseconds"))
            return 1_000;
        else if (units.equals("nanoseconds"))
            return 1;
        else throw new IllegalArgumentException("Invalid time units: " + units);
    }

    private Long start;

    private final int timeFactor;
}
